### PR TITLE
Haystack

### PR DESCRIPTION
**Pull request by r0man**
This PR adds the Haystack stacktrace analyzer and parser, originally part of cider-nrepl and Orchard.

I checked that inlining of haystack works in cider-nrepl.

Before submitting a PR make sure the following things have been done (and denote this
by checking the relevant checkboxes):

- [x] The commits are consistent with our [contribution guidelines](CONTRIBUTING.md)
- [x] You've added tests (if possible) to cover your change(s)
- [x] All tests are passing (run `lein do clean, test`)
- [x] Code inlining with mranderson works and tests pass with inlined code (run `./build.sh install` -- takes a long time)
- [x] You've updated the changelog (if adding/changing user-visible functionality)
- [x] You've updated the readme (if adding/changing user-visible functionality)

Thanks!

### CODE REVIEW

**Comment by r0man** [comment]
Hi @bbatsov @vemv ,

I'm back from vacation.

Here's the PR with the stacktrace parsers and the analyzer which I moved from cider-nrepl to this repo.

A couple of notes:
- Since Instaparse works as well in ClojureScript, I changed the parsers to cljc files, so they work on both platforms.
- You can run the ClojureScript tests with Node installed by running `lein test-cljs`. @vemv If we want this on CI can you help with setting this up?
- Regarding the analyzer. I think it fits ito this repo, bu it's the only part that uses non-trival functionality from Orchard. So haystack depends on Orchard. Not sure if this is an issue.

WDYT?

**Comment by bbatsov** [comment]
> Regarding the analyzer. I think it fits ito this repo, bu it's the only part that uses non-trival functionality from Orchard. So haystack depends on Orchard. Not sure if this is an issue.

Ideally it shouldn't. What does it use from Orchard exactly?

**Comment by r0man** [comment]
Orchard's Info system, resolving Java symbols, resource loading and namespace info:
https://github.com/r0man/haystack/blob/haystack/src/haystack/analyzer.clj#L11-L14

**Comment by r0man** [comment]
Looks like the tests did not get triggered on CI. Any ideas?

**Comment by bbatsov** [comment]
> Orchard's Info system, resolving Java symbols, resource loading and namespace info:

I had forgotten about this by now. That'd be one dependency we'll have a hard time removing, so I guess we should go with it for now.

**Comment by bbatsov** [comment]
I've enabled the CI build and I think we can merge this at this point, so @r0man can work towards the cider-nrepl integration. We can always make small improvements here as we go.

File: project.clj [hunk 1]
```diff
@@ -33,14 +46,16 @@
              :cljfmt [:test
                       {:plugins [[lein-cljfmt "0.9.0" :exclusions [org.clojure/clojure
                                                                    org.clojure/clojurescript]]]}]
-             :eastwood {:plugins         [[jonase/eastwood "1.3.0"]]
+             :eastwood {:plugins [[jonase/eastwood "1.3.0"]]
                         :eastwood {;; :implicit-dependencies would fail spuriously when the CI matrix runs for Clojure < 1.10,
                                    ;; because :implicit-dependencies can only work for a certain corner case starting from 1.10.
                                    :exclude-linters [:implicit-dependencies]
                                    :exclude-namespaces [refactor-nrepl.plugin]
-                                   :add-linters [:performance :boxed-math]
```
[the remark this answers is not in the document]
Comment (r0man) [reply project.clj:41]: I opened a PR for this: https://github.com/clojure-emacs/haystack/pull/4

File: .clj-kondo/config.edn [hunk 2]
```diff
@@ -1,10 +1,11 @@
-{:linters {:discouraged-var                       {clojure.core/read-string {:message "Please prefer clojure.edn/read-string"}}
+{:lint-as {instaparse.core/defparser clojure.core/def}
+ :linters {:discouraged-var                       {clojure.core/read-string {:message "Please prefer clojure.edn/read-string"}}
+           :unresolved-var                        {:exclude [instaparse.core/transform]}
            ;; Enable some disabled-by-default linters.
            :docstring-leading-trailing-whitespace {:level :warning}
            :keyword-binding                       {:level :warning}
            :reduce-without-init                   {:level :warning}
            :redundant-fn-wrapper                  {:level :warning}
-           :shadowed-var                          {:level :warning}
```
[the remark this answers is not in the document]
Comment (r0man) [reply .clj-kondo/config.edn:7]: This one gives a ton of warnings, for my opinion reasonable argument and var names:

```
src/haystack/analyzer.clj:32:10: warning: Shadowed var: clojure.core/class
src/haystack/analyzer.clj:47:4: warning: Shadowed var: clojure.core/class
src/haystack/analyzer.clj:73:12: warning: Shadowed var: clojure.core/type
src/haystack/analyzer.clj:73:17: warning: Shadowed var: clojure.core/class
src/haystack/analyzer.clj:76:12: warning: Shadowed var: clojure.core/ns
src/haystack/analyzer.clj:76:15: warning: Shadowed var: clojure.core/fn
src/haystack/analyzer.clj:79:11: warning: Shadowed var: clojure.core/fn
src/haystack/analyzer.clj:100:9: warning: Shadowed var: clojure.core/type
src/haystack/analyzer.clj:194:23: warning: Shadowed var: clojure.core/ns
src/haystack/analyzer.clj:236:12: warning: Shadowed var: clojure.core/class
src/haystack/analyzer.clj:294:33: warning: Shadowed var: clojure.core/val
src/haystack/parser/clojure/repl.cljc:67:35: warning: Shadowed var: clojure.core/type
src/haystack/parser/clojure/repl.cljc:67:35: warning: Shadowed var: cljs.core/type
src/haystack/parser/clojure/stacktrace.cljc:69:35: warning: Shadowed var: clojure.core/type
src/haystack/parser/clojure/stacktrace.cljc:69:35: warning: Shadowed var: cljs.core/type
src/haystack/parser/clojure/tagged_literal.cljc:20:12: warning: Shadowed var: clojure.core/class
src/haystack/parser/java.cljc:66:35: warning: Shadowed var: clojure.core/type
src/haystack/parser/java.cljc:66:35: warning: Shadowed var: cljs.core/type
test/haystack/analyzer_test.clj:14:22: warning: Shadowed var: clojure.core/name
test/haystack/analyzer_test.clj:17:26: warning: Shadowed var: clojure.core/name
test/haystack/analyzer_test.clj:186:9: warning: Shadowed var: clojure.core/clojure-version
test/haystack/analyzer_test.clj:275:9: warning: Shadowed var: haystack.analyzer-test/causes
test/haystack/analyzer_test.clj:278:21: warning: Shadowed var: clojure.core/class
test/haystack/analyzer_test.clj:303:21: warning: Shadowed var: clojure.core/class
test/haystack/analyzer_test.clj:313:21: warning: Shadowed var: clojure.core/class
test/haystack/analyzer_test.clj:324:9: warning: Shadowed var: haystack.analyzer-test/causes
test/haystack/analyzer_test.clj:327:21: warning: Shadowed var: clojure.core/class
test/haystack/analyzer_test.clj:355:21: warning: Shadowed var: clojure.core/class
test/haystack/analyzer_test.clj:374:21: warning: Shadowed var: clojure.core/class
test/haystack/analyzer_test.clj:394:9: warning: Shadowed var: haystack.analyzer-test/causes
test/haystack/analyzer_test.clj:397:21: warning: Shadowed var: clojure.core/class
test/haystack/analyzer_test.clj:417:9: warning: Shadowed var: haystack.analyzer-test/causes
test/haystack/analyzer_test.clj:420:21: warning: Shadowed var: clojure.core/class
test/haystack/analyzer_test.clj:448:21: warning: Shadowed var: clojure.core/class
test/haystack/analyzer_test.clj:476:21: warning: Shadowed var: clojure.core/class
test/haystack/analyzer_test.clj:505:9: warning: Shadowed var: haystack.analyzer-test/causes
test/haystack/analyzer_test.clj:530:21: warning: Shadowed var: clojure.core/class
test/haystack/analyzer_test.clj:558:21: warning: Shadowed var: clojure.core/class
test/haystack/analyzer_test.clj:577:21: warning: Shadowed var: clojure.core/class
test/haystack/parser/aviso_test.cljc:21:39: warning: Shadowed var: clojure.core/type
test/haystack/parser/aviso_test.cljc:21:39: warning: Shadowed var: cljs.core/type
test/haystack/parser/aviso_test.cljc:27:39: warning: Shadowed var: clojure.core/type
test/haystack/parser/aviso_test.cljc:27:39: warning: Shadowed var: cljs.core/type
test/haystack/parser/aviso_test.cljc:33:39: warning: Shadowed var: clojure.core/type
test/haystack/parser/aviso_test.cljc:33:39: warning: Shadowed var: cljs.core/type
test/haystack/parser/aviso_test.cljc:57:39: warning: Shadowed var: clojure.core/type
test/haystack/parser/aviso_test.cljc:57:39: warning: Shadowed var: cljs.core/type
test/haystack/parser/aviso_test.cljc:63:39: warning: Shadowed var: clojure.core/type
test/haystack/parser/aviso_test.cljc:63:39: warning: Shadowed var: cljs.core/type
test/haystack/parser/aviso_test.cljc:69:39: warning: Shadowed var: clojure.core/type
test/haystack/parser/aviso_test.cljc:69:39: warning: Shadowed var: cljs.core/type
test/haystack/parser/aviso_test.cljc:93:39: warning: Shadowed var: clojure.core/type
test/haystack/parser/aviso_test.cljc:93:39: warning: Shadowed var: cljs.core/type
test/haystack/parser/aviso_test.cljc:119:39: warning: Shadowed var: clojure.core/type
test/haystack/parser/aviso_test.cljc:119:39: warning: Shadowed var: cljs.core/type
test/haystack/parser/aviso_test.cljc:134:39: warning: Shadowed var: clojure.core/type
test/haystack/parser/aviso_test.cljc:134:39: warning: Shadowed var: cljs.core/type
test/haystack/parser/aviso_test.cljc:153:31: warning: Shadowed var: clojure.core/type
test/haystack/parser/aviso_test.cljc:153:31: warning: Shadowed var: cljs.core/type
test/haystack/parser/clojure/repl_test.cljc:21:45: warning: Shadowed var: clojure.core/type
test/haystack/parser/clojure/repl_test.cljc:21:45: warning: Shadowed var: cljs.core/type
test/haystack/parser/clojure/repl_test.cljc:31:45: warning: Shadowed var: clojure.core/type
test/haystack/parser/clojure/repl_test.cljc:31:45: warning: Shadowed var: cljs.core/type
test/haystack/parser/clojure/repl_test.cljc:41:45: warning: Shadowed var: clojure.core/type
test/haystack/parser/clojure/repl_test.cljc:41:45: warning: Shadowed var: cljs.core/type
test/haystack/parser/clojure/repl_test.cljc:69:39: warning: Shadowed var: clojure.core/type
test/haystack/parser/clojure/repl_test.cljc:69:39: warning: Shadowed var: cljs.core/type
test/haystack/parser/clojure/repl_test.cljc:93:39: warning: Shadowed var: clojure.core/type
test/haystack/parser/clojure/repl_test.cljc:93:39: warning: Shadowed var: cljs.core/type
test/haystack/parser/clojure/repl_test.cljc:108:39: warning: Shadowed var: clojure.core/type
test/haystack/parser/clojure/repl_test.cljc:108:39: warning: Shadowed var: cljs.core/type
test/haystack/parser/clojure/repl_test.cljc:125:31: warning: Shadowed var: clojure.core/type
test/haystack/parser/clojure/repl_test.cljc:125:31: warning: Shadowed var: cljs.core/type
test/haystack/parser/clojure/stacktrace_test.cljc:21:45: warning: Shadowed var: clojure.core/type
test/haystack/parser/clojure/stacktrace_test.cljc:21:45: warning: Shadowed var: cljs.core/type
test/haystack/parser/clojure/stacktrace_test.cljc:31:45: warning: Shadowed var: clojure.core/type
test/haystack/parser/clojure/stacktrace_test.cljc:31:45: warning: Shadowed var: cljs.core/type
test/haystack/parser/clojure/stacktrace_test.cljc:41:45: warning: Shadowed var: clojure.core/type
test/haystack/parser/clojure/stacktrace_test.cljc:41:45: warning: Shadowed var: cljs.core/type
test/haystack/parser/clojure/stacktrace_test.cljc:71:39: warning: Shadowed var: clojure.core/type
test/haystack/parser/clojure/stacktrace_test.cljc:71:39: warning: Shadowed var: cljs.core/type
test/haystack/parser/clojure/stacktrace_test.cljc:95:39: warning: Shadowed var: clojure.core/type
test/haystack/parser/clojure/stacktrace_test.cljc:95:39: warning: Shadowed var: cljs.core/type
test/haystack/parser/clojure/stacktrace_test.cljc:110:39: warning: Shadowed var: clojure.core/type
test/haystack/parser/clojure/stacktrace_test.cljc:110:39: warning: Shadowed var: cljs.core/type
test/haystack/parser/clojure/stacktrace_test.cljc:127:31: warning: Shadowed var: clojure.core/type
test/haystack/parser/clojure/stacktrace_test.cljc:127:31: warning: Shadowed var: cljs.core/type
test/haystack/parser/clojure/tagged_literal_test.cljc:21:39: warning: Shadowed var: clojure.core/type
test/haystack/parser/clojure/tagged_literal_test.cljc:21:39: warning: Shadowed var: cljs.core/type
test/haystack/parser/clojure/tagged_literal_test.cljc:27:39: warning: Shadowed var: clojure.core/type
test/haystack/parser/clojure/tagged_literal_test.cljc:27:39: warning: Shadowed var: cljs.core/type
test/haystack/parser/clojure/tagged_literal_test.cljc:33:39: warning: Shadowed var: clojure.core/type
test/haystack/parser/clojure/tagged_literal_test.cljc:33:39: warning: Shadowed var: cljs.core/type
test/haystack/parser/clojure/tagged_literal_test.cljc:57:39: warning: Shadowed var: clojure.core/type
test/haystack/parser/clojure/tagged_literal_test.cljc:57:39: warning: Shadowed var: cljs.core/type
test/haystack/parser/clojure/tagged_literal_test.cljc:81:39: warning: Shadowed var: clojure.core/type
test/haystack/parser/clojure/tagged_literal_test.cljc:81:39: warning: Shadowed var: cljs.core/type
test/haystack/parser/clojure/tagged_literal_test.cljc:108:31: warning: Shadowed var: clojure.core/type
test/haystack/parser/clojure/tagged_literal_test.cljc:108:31: warning: Shadowed var: cljs.core/type
test/haystack/parser/clojure/tagged_literal_test.cljc:115:31: warning: Shadowed var: clojure.core/type
test/haystack/parser/clojure/tagged_literal_test.cljc:115:31: warning: Shadowed var: cljs.core/type
test/haystack/parser/clojure/throwable_test.cljc:29:36: warning: Shadowed var: clojure.core/type
test/haystack/parser/clojure/throwable_test.cljc:29:36: warning: Shadowed var: cljs.core/type
test/haystack/parser/clojure/throwable_test.cljc:35:36: warning: Shadowed var: clojure.core/type
test/haystack/parser/clojure/throwable_test.cljc:35:36: warning: Shadowed var: cljs.core/type
test/haystack/parser/clojure/throwable_test.cljc:41:36: warning: Shadowed var: clojure.core/type
test/haystack/parser/clojure/throwable_test.cljc:41:36: warning: Shadowed var: cljs.core/type
test/haystack/parser/clojure/throwable_test.cljc:64:47: warning: Shadowed var: clojure.core/type
test/haystack/parser/clojure/throwable_test.cljc:84:36: warning: Shadowed var: clojure.core/type
test/haystack/parser/clojure/throwable_test.cljc:84:36: warning: Shadowed var: cljs.core/type
test/haystack/parser/java_test.cljc:21:45: warning: Shadowed var: clojure.core/type
test/haystack/parser/java_test.cljc:21:45: warning: Shadowed var: cljs.core/type
test/haystack/parser/java_test.cljc:31:45: warning: Shadowed var: clojure.core/type
test/haystack/parser/java_test.cljc:31:45: warning: Shadowed var: cljs.core/type
test/haystack/parser/java_test.cljc:41:45: warning: Shadowed var: clojure.core/type
test/haystack/parser/java_test.cljc:41:45: warning: Shadowed var: cljs.core/type
test/haystack/parser/java_test.cljc:71:39: warning: Shadowed var: clojure.core/type
test/haystack/parser/java_test.cljc:71:39: warning: Shadowed var: cljs.core/type
test/haystack/parser/java_test.cljc:95:39: warning: Shadowed var: clojure.core/type
test/haystack/parser/java_test.cljc:95:39: warning: Shadowed var: cljs.core/type
test/haystack/parser/java_test.cljc:110:39: warning: Shadowed var: clojure.core/type
test/haystack/parser/java_test.cljc:110:39: warning: Shadowed var: cljs.core/type
test/haystack/parser/java_test.cljc:127:31: warning: Shadowed var: clojure.core/type
test/haystack/parser/java_test.cljc:127:31: warning: Shadowed var: cljs.core/type
test/haystack/parser/test.cljc:13:12: warning: Shadowed var: clojure.core/name
test/haystack/parser/test.cljc:24:10: warning: Shadowed var: clojure.core/class
```

I don't really like that linter. :) If you really want it, I can go and look up some synonyms for all those vars.

File: eastwood.clj [hunk 3]
```diff
@@ -0,0 +1 @@
+
```
[the remark this answers is not in the document]
Comment (r0man) [reply eastwood.clj:1]: Fixed, in https://github.com/clojure-emacs/haystack/pull/5

**Pull request merged**